### ISSUE TITLE
Korean hierarchy appears to be incorrect

### ISSUE THREAD

**Issue by ianthetechie** [whosonfirst-data/whosonfirst-data#2141]
I have been doing some work with WOF data recently and noticed the admin hierarchies in Korea are a bit convoluted.

# TL;DR summary

1. All districts of Seoul are (in my understanding) incorrectly classified as counties rather than boroughs.
2. Most neighborhoods in Seoul is incorrectly classified as a child of [Yongsan](https://spelunker.whosonfirst.org/id/890475295/).

# Details

Seoul is both a special region (aka "province" in some taxonomies) and a city (locality), and as far as I can tell, this is correctly classified. Within Seoul, there are a bunch of administrative units, which I believe translate best to boroughs in the WOF taxonomy. Further each district (borough) is subdivided further into what translate best as neighborhoods.

Yongsan is one of these districts (borough), and happens to be near the center. For some reason, Yongsan and other districts I've looked up have been classified as counties. For example, see [Songpa](https://spelunker.whosonfirst.org/id/890473255/). When you look at Songpa, you'll notice that it is a descendent of the Seoul region, which is correct. However, it probably should be a direct descendent of Seoul the locality.

Taking an example, the neighborhood in which I live, [Jamsil-dong](https://spelunker.whosonfirst.org/id/85926401/), is incorrectly classified as a child of Yongsan, when it is actually in [Songpa](https://spelunker.whosonfirst.org/id/890473255/), both legally and geographically.

(Aside: it's a bit inconsistent that the neighborhoods have a Korean suffix, but boroughs don't; I'll have to double check official guidance on that but it's quite common to include the suffix, like Songpa-gu in English signage locally.)

I would be happy to contribute PRs / QA any fixes if anyone can point me in the right direction. It seems like this is a somewhat fragile operation and there isn't any formalized tooling available yet per the README, but I'm happy to take a crack with some direction.

**Comment by nvkelso** [comment]
Thanks for the detailed report and recommended fixes! It sounds tractable, and a little like New York City -- so we have a good pattern to follow. Yes. PRs are gladly accepted 😄 

@stepps00 lets review Ian's proposal, confirm plan, and link to relevant PRs to pattern off? Since we're mostly modifying neighbourhoods we can probably hand curate the admin hierarchies versus PIP (point-in-polygon) after?

**Comment by ianthetechie** [comment]
@justinelliotmeyers indeed I am; the reason I noticed the issue is because I live here. I'm vaguely aware of their open geographic data request procedure, but do they actually release data we could use in a format easily ingestible into WOF with the relationships in place? I'm a bit cynical about certain quality standards when a govt agency is translating into an unfamiliar format, but happy to give that a shot.

**Comment by justinelliotmeyers** [comment]
@ianthetechie  Free federal data is here: https://sgis.kostat.go.kr/view/pss/openDataIntrcn

The site also has terms of use/ copyright pages as well (*which I believe the data is open) But always good to have a second set of eyes on it, especially if in a different language. WOF can ingest shapefile data, so if you could formally request data down to dong/ ri level (*although I don't think census formally has ri data but not 100% certain on that) and then the handful of parent layers, that would be great. Everything will stack nicely for a parent child structure here. And data should come in Korean and English 

I can get into more details if you'd like about other build types with informal south korea data and a way more in depth overall description. But this is a great start if you can assist.

**Comment by justinelliotmeyers** [comment]
@ianthetechie just wanted to follow-up and see if you can grab a few datasets mentioned above. If you can't let me know and I can ping a few other colleagues. Appreciate your time and help! Have a good weekend
Justin

**Comment by justinelliotmeyers** [comment]
@ianthetechie I just rebuilt South Korea - data from June 2023. Confirmed Jamsil-dong is a child of Songpa-gu under Seoul.

**Comment by ianthetechie** [comment]
Quick bump out of curiosity if there are any blockers on getting this updated, since it sounds like it worked ;) Also happy to review a PR for accuracy

**Comment by nvkelso** [comment]
@stepps00 Want to tackle this in your spare time?

@ianthetechie tl;dr everyone got sucked into other projects versus working on WOF 9-5 so the pace has slowed down a bit.

**Comment by ianthetechie** [comment]
No worries @nvkelso. I'd be happy to help; just not sure the steps that need to be taken 😅

**Comment by stepps00** [comment]
Thanks for all of the info here.. here's a screenshot of existing coverage around Seoul. Localities are outlined in dark blue. Counties are lighter blue, with the counties ([districts](https://en.wikipedia.org/wiki/List_of_districts_of_Seoul#List_by_population_and_area), really) within Seoul highlighted:

<img width="1384" alt="Screenshot 2024-11-13 at 10 34 31 PM" src="https://github.com/user-attachments/assets/e4e36934-376c-42a0-a4cf-bff386927254">

Based on these comments, I'm thinking the updates would be:

- Update the locality of Seoul's geometry to match the geometry of all combined counties/districts within Seoul
- Supersede the existing counties within Seoul into new borough records
- Create a new county record for Seoul
- re-PIP any updated records to update hierarchies and parent_ids for all records

That would result in:

- 166 neighbourhoods of Seoul
- 1 locality of Seoul
- 25 boroughs of Seoul to represent districts
- 1 county of Seoul
- 1 region of Seoul

This would also solve the issue of all neighbourhoods being parented by Yongsan; because the locality of Seoul is parented by a single county today, that single county is tricklded down to the hierarchy of all descendant neighbourhoods.

For what it's worth, here is the current neighbourhood coverage:

<img width="1381" alt="Screenshot 2024-11-13 at 10 45 08 PM" src="https://github.com/user-attachments/assets/8b164c81-ae47-4c55-a873-3b69e834d4be">

Going forward, any suggestions on how to request official data from https://sgis.kostat.go.kr/view/pss/openDataIntrcn ? I'm directed to a login page when trying to access data, and the second link (http://data.nsdi.go.kr/dataset/20180927ds0059) seems to be dead now. I can take another look at new data sources when I start updates.

@ianthetechie I'd say the biggest help here would be accessing new geometries and helping with QA once a pull request is up, if possible. Let me know if this proposal makes sense to you.

**Comment by nvkelso** [comment]
+1 to the proposed plan. Makes sense to cleanup existing WOF records first and then if new data arrives 2nd PR.

**Comment by ianthetechie** [comment]
+1 to the proposed plan from me as well. Ping me on the PR when it's up and I can help sanity check that the hierarchies and geometries look right :) (Is there a way I could like run Spelunker or a similar tool locally? Or is it just throwing GeoJSON in QGIS?).

On the data request, I am also seeing a login page, using I-PIN (a service I've never used but am vaguely aware of) and cell phone authentication (would not be accessible unless you have a non-temp cell phone linked to your national ID). I'm mucking around with it real quick to see if I can find the relevant data. Do you happen to know what you downloaded before?

**Comment by ianthetechie** [comment]
OK, TL;DR...

1. You need an account (either I-PIN or register with a Korean cell phone tied to your ID)
2. After about 12 steps (typical), you get an account. The data portal for official boundaries lives here: https://sgis.kostat.go.kr/statexp/view/index#. You can use this to get shapefiles. Also their SSL cert expired like a year ago :D 
  - The site is only available in Korean, but you need to download shapefiles for all types of boundaries for the entire country (잔국).

<img width="354" alt="image" src="https://github.com/user-attachments/assets/90562a22-531a-4abb-a018-7544e1478f34">

Download button is at the bottom of the form. When you click the download button, it may take a minute or more to fetch your file (pop up a web inspector if you're curious). The text in the shapefiles is encoded in EUC-KR; QGIS automatic detection failed.

Where can I dump the shapefiles for processing (or maybe I can take a whack at the next update cycle)? I've got all the 2024 data.

**Comment by nvkelso** [comment]
What about via Google drive?

**Comment by justinelliotmeyers** [comment]
I believe the kostat data is the 행정동 version of Korea's boundaries. These are mainly for government purposes. I believe WOF would want the 법정동 boundaries. Which makes sense to people who live there. Different boundaries for different purposes. And companies in South Korea want both supported - which is possible, just a fun game when supporting in a type in search capacity ;)

**Comment by ianthetechie** [comment]
Hmm good point. Looking closely, yeah, these are the administrative boundaries for things like "which local govt office / polling place do you need to go to." But nobody would actually say they live in, say, 방화1동.

<img width="876" alt="image" src="https://github.com/user-attachments/assets/fc809702-7ee7-4c96-8578-c8ffda29d129">

The maps on juso.go.kr have the more logical boundaries that most people would reference (I guess these would be "neighborhoods" rather than administrative subdivisions in WOF taxonomy), but I can't find a way to download these 😅 

<img width="931" alt="image" src="https://github.com/user-attachments/assets/03abb293-3cc1-413f-b23b-d862d126e584">

Any ideas where to get these?

> Different boundaries for different purposes. And companies in South Korea want both supported - which is possible, just a fun game when supporting in a type in search capacity ;)

Yeah 😅 I feel this is a never ending problem haha. Like "postal cities" you'd put on an envelope to be mailed. They diverge from administrative reality like... a LOT. Which matters for you depends a lot on your use case...

At any rate, here's a folder with the shapefiles I have so far. If I understand the WOF taxonomy correctly, all of these are useful, but the 읍/면/동 boundaries (specifically 동) is a purely administrative boundary. Geocoders would be more interested in neighborhoods, which are not included. https://drive.google.com/drive/folders/1WkdUhbiuuuLn_-vXBn6LR3WYnfXJeVBw?usp=sharing

**Comment by ianthetechie** [comment]
This is quite an interesting rabbit hole btw... interestingly, OSM also mostly concerns itself with these administrative boundaries rather than the legal ones; OSM coverage of Seoul is fantastic in the administrative sense, but when it comes to the legal ones, I spot checked a few this morning that I'm familiar with, and it ranged from nonexistent to point-only to full relation.

**Comment by justinelliotmeyers** [comment]
I probably shouldn't have said anything! 

I just didn't want these to be plopped in, then anyone realize ri (li) village data doesn't really roll up to them, or places were missing. Then creating the fix is... honestly a lot of work.

I maintained both versions for Samsung and LG for several years. Lots of updates. Lots of games. Dokdo and a few islands along the North Korean border have fun codes.

the governments open data portal should have admin 0,1,2,3,4. The second trick is the license.

There are a few unofficial layers out there that are 99% identical to the govt data if this gets stuck.
